Fix: use Zod validation for JSON requests (supersedes #385)

ayewo opened the pull request:
This PR uses a separate `fix/zod-v2` branch name rather than `main` and supersedes the one I opened in #385.

Closes #384 so I can /claim #384

ayewo commented:
Please also see my comment regarding `.passthrough()` here: https://github.com/Cap-go/capgo/commit/5720fccbb2b5364f3a09fd86c87987fd1cebbc6c#r128555567

riderx commented:
Thanks @ayewo can we have to backend test for this ?

WcaleNieWolny commented on supabase/functions/updates/index.ts:
```diff
@@ -3,27 +3,42 @@ import { serve } from 'https://deno.land/std@0.200.0/http/server.ts'
+  }),
   is_emulator: z.boolean().default(false),
   is_prod: z.boolean().default(true),
-}).passthrough()
```
Without passtrough all data not defined in zod schema will be stripped away

WcaleNieWolny commented on supabase/functions/updates/index.ts:
```diff
@@ -33,6 +48,13 @@ const headersSchema = z.object({
 async function main(_url: URL, _headers: BaseHeaders, _method: string, body: AppInfos) {
+  let parseResult = null
+  try {
+    parseResult = jsonRequestSchema.parse(body)
```
You can use safaparse here

WcaleNieWolny commented on supabase/functions/stats/index.ts:
```diff
@@ -13,9 +14,54 @@ const failActions = [
 async function main(url: URL, headers: BaseHeaders, method: string, body: AppStats) {
   try {
     console.log('body', body)
+    let _parseResult = null
```
Why is this unused? The entire idea is to add type safety and if you do not use zod result you do not have type safety

ayewo replied:
Fixed in the latest [commit 89a985](https://github.com/Cap-go/capgo/commit/89a985b8f1ec926269aed3fa5f7134b7f8afca6e).

WcaleNieWolny commented on supabase/functions/stats/index.ts:
```diff
@@ -13,9 +14,54 @@ const failActions = [
     console.log('body', body)
+    let _parseResult = null
+    try {
+      _parseResult = jsonRequestSchema.parse(body)
```
You can use safeparse

ayewo replied:
Switched to `safeParse` in the latest commit.

WcaleNieWolny commented on supabase/functions/stats/index.ts:
```diff
@@ -13,9 +14,54 @@ const failActions = [
+      _parseResult = jsonRequestSchema.parse(body)
+    }
+    catch (error) {
+      return sendRes({ error: `Cannot parse json: ${error}` }, 400)
```
This error response will look like unreadable garbage, zod errors are not great to parse to json as far as I remember from my testing. Could you provide an example of a failed parse request?

ayewo replied:
A failed parse typically looks like this in the console:
```bash
serving the request with /home/deno/functions/updates
[Info] Listening on http://localhost:9999/

[Error] sendRes error {
  "error": "Cannot parse json: [\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"string\",\n    \"received\": \"undefined\",\n    \"path\": [\n      \"app_id\"\n    ],\n    \"message\": \"App ID is required\"\n  }\n]"
}

[Error] sendRes error "{\"error\":\"Cannot parse json: [\\n  {\\n    \\\"code\\\": \\\"invalid_type\\\",\\n    \\\"expected\\\": \\\"string\\\",\\n    \\\"received\\\": \\\"undefined\\\",\\n    \\\"path\\\": [\\n      \\\"app_id\\\"\\n    ],\\n    \\\"message\\\": \\\"App ID is required\\\"\\n  }\\n]\"}"
```

Since the intent behind the use of Zod validation is to mitigate abuse from bad actors as per https://github.com/Cap-go/capgo/issues/384 this response will serve its purpose (by slowing them down). 

An unmodified Capgo plugin should still be able to consume the `/update` and `/stats` endpoints as before so there should be no impact to existing users.

ayewo commented:
> Thanks @ayewo can we have to backend test for this ?

You specifically requested Zod validation and did not mention any kind of (integration) tests.

So, I would really appreciate it if you can please increase the payment.

ayewo commented:
> Thanks @ayewo can we have to backend test for this ?

@riderx You specifically requested Zod validation and did not mention any kind of (integration) tests.

So, I would really appreciate it if you can please increase the payment.

WcaleNieWolny commented:
@ayewo I have implemented the fixes I requested + tests for the `updates` endpoint. I have not wrote tests for the `stats` endpoint as it currently is untested thus it would require testing the entire endpoint and this is not the goal of this PR

The patch is available [here](https://gist.github.com/WcaleNieWolny/e42b9046b4d921ead31f6fdb0b7b764a)

Please be aware that I am not doing this to get the increased payment you requested. I am doing that so that this PR can be merged.

When writing this I discovered that the `passtrough` is not needed anymore (since you added the full schema and just a part of the schema as I did) + that this PR would not build in the current step.
![image](https://github.com/Cap-go/capgo/assets/50914789/add29455-1aa7-48a3-8532-6f4df6972f89)
As for safeparse I have no idea what are you talking about. I used safeparse + wrote tests for this and this works just fine.

If you do not want to use my tests that is fine with me, but please use the general fixes to the endpoints

ayewo commented:
> When writing this I discovered that the `passtrough` is not needed anymore (since you added the full schema and just a part of the schema as I did) + that this PR would not build in the current step. 

1. I tried to point this out [earlier](https://github.com/Cap-go/capgo/commit/5720fccbb2b5364f3a09fd86c87987fd1cebbc6c#r128555567) but you [doubled-down](https://github.com/Cap-go/capgo/pull/399#discussion_r1339581908) on your position. Good that you've come to the same conclusion after you tested it yourself.

![image](https://user-images.githubusercontent.com/50914789/271810996-add29455-1aa7-48a3-8532-6f4df6972f89.png) 

2. Thanks for catching this 👍. I could have sworn I had removed those lines already.


> As for safeparse I have no idea what are you talking about. I used safeparse + wrote tests for this and this works just fine.

3. Again, instead of a back and forth, I felt it would be better if I demonstrated the issue I was describing using tests. In my testing `safeParse()` behaved differently from `parse()` which is why I switched it out.

WcaleNieWolny commented:
> I tried to point this out [earlier](https://github.com/Cap-go/capgo/commit/5720fccbb2b5364f3a09fd86c87987fd1cebbc6c#r128555567) but you https://github.com/Cap-go/capgo/pull/399#discussion_r1339581908 on your position

ayewo commented:
> PS: I am really sorry about closing and reopening your PR, my pc sometimes does unexplainable things without any input. I take full responsibility for this

No worries, it's not an issue at all. (The *"Close with comment"* button is right next to the *"Comment"* button so mistakes are bound to happen.)


I've been meaning to ask: in the patch you [shared](https://gist.github.com/WcaleNieWolny/e42b9046b4d921ead31f6fdb0b7b764a), the tests are shorter now (and easier to follow 😀) but I'm curious as to why you didn't use popular library like `jest` / [`ts-jest`](https://github.com/kulshekhar/ts-jest) for the tests?

Use of `jest` would make the tests easier to maintain. Specifically, it will make it easy to:
- group related tests using `describe('<describe-title>', () => {})`; 
- give each test a title via the `test('<test-title>',() => {})` method;
- reduce duplication by putting common setup and teardown actions inside `beforeEach()` and `afterEach()` methods;
- produce a nice summary of how many tests passed or failed (and the time they took).

FWIW, when I attempted to add tests for my changes last week, some of the tests failed out of the box. 
Because the tests lacked titles and the other items I mentioned above, I had no idea whether it was an easy fix or it would quickly turn into an hour long debugging session[^1] so I had to abandon it. 


[^1]: Assuming there was some indication of how many tests passed or failed e.g. 1 failed test out of 20, I could have tracked down the failing test and changed it from `test()` to `test.skip()` to skip it so the rest of the tests could pass.

WcaleNieWolny commented:
> why you didn't use popular library like jest / [ts-jest](https://github.com/kulshekhar/ts-jest) for the tests?

I do not know, really. I have no idea, the more I develop the custom testing lib the more "clean" it becomes tho I do wonder why am I developing this. If I would to remove this custom framework I would have use `playwright` as it is already integrated into the project.

> FWIW, when I attempted to add tests for my changes last week, some of the tests failed out of the box.

Some will. The backend tests are stateful (yes, really. To run them do `supabase db reset`) and the CLI tests require ENV variables. I do agree that the backend (`/updates` endpoint tests) are a hot mess and I was planning on cleaning this up after I do [major or minor for security](https://github.com/Cap-go/capgo/issues/360). I have implemented more and more of the logic required for tests into playwright and I was considering moving 100% to playwright. Right now this is just a thought. I might keep the custom solution. I do not know.

I am sorry you had such a horrible experience with the tests. The entire thing is still very much WIP and I am looking forward on hearing some ideas on how we can improve it

WcaleNieWolny commented:
> the tests are shorter now

Yes they are, I improved them but I did not want to base my patch on a branch not yet merged into main at the time

ayewo commented:
@WcaleNieWolny 
> If you experienced something like this then it is likely a bug in zod. I have looked at the documentations and it says:
...
> There is no mention about any difference in the behavior. If you have steps to reproduce you could try to report it to zod.

I've tested again and you are right. Turned out to be user error on my part. Thanks for the nudge.

WcaleNieWolny reviewed (approved):
Overall LGTM 🚀 

As martin suggested: Tests would be nice

ayewo commented:
> Overall LGTM 🚀
> 
> As martin suggested: Tests would be nice

Of course! I've written them and planned to include them in my next commit to this branch, but unfortunately they no longer run after I rebased on `main` ... due to some TypeScript error from an unrelated file (that I didn't touch).
<details>
<pre>
error: TS2769 [ERROR]: No overload matches this call.
  Overload 1 of 2, '(values: { app_id: string; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id: string; is_emulator?: boolean | null | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }, options?: { ...; } | undefined): PostgrestFilterBuilder<...>', gave the following error.
    Argument of type '{ app_id?: string | undefined; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id?: string | undefined; is_emulator?: boolean | ... 1 more ... | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }' is not assignable to parameter of type '{ app_id: string; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id: string; is_emulator?: boolean | null | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }'.
      Types of property 'app_id' are incompatible.
        Type 'string | undefined' is not assignable to type 'string'.
          Type 'undefined' is not assignable to type 'string'.  Overload 2 of 2, '(values: { app_id: string; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id: string; is_emulator?: boolean | null | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }[], options?: { ...; } | undefined): PostgrestFilterBuilder<...>', gave the following error.
    Argument of type '{ app_id?: string | undefined; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id?: string | undefined; is_emulator?: boolean | ... 1 more ... | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }' is not assignable to parameter of type '{ app_id: string; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id: string; is_emulator?: boolean | null | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }[]'.
      Type '{ app_id?: string | undefined; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id?: string | undefined; is_emulator?: boolean | ... 1 more ... | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }' is missing the following properties from type '{ app_id: string; created_at?: string | null | undefined; custom_id?: string | undefined; date_id?: string | null | undefined; device_id: string; is_emulator?: boolean | null | undefined; ... 7 more ...; version_build?: string | ... 1 more ... | undefined; }[]': length, pop, push, concat, and 35 more.
    .upsert(device), sendDeviceToTinybird([device])])
            ~~~~~~
</pre>
</details>

ayewo commented on supabase/functions/stats/index.ts:
```diff
@@ -13,9 +14,54 @@ const failActions = [
+    required_error: 'Device ID is required',
+    invalid_type_error: 'Device ID must be a string',
+  }).max(36),
+  platform: z.string({
```
Somehow I missed this. I tried a union but it didn't work.

WcaleNieWolny commented:
@riderx please fix this tinibird issue when possible

@ayewo comment this out for now if possible, martin will likely fix this as some point as he is responsible for this

WcaleNieWolny commented:
Also the way you run the tests is kind of wierd. Why `deno test --allow-all supabase/functions/**/index.test.ts
`? Have you wrote this as a unit test and if so have you integrated this with the current testing framework (the one tiggered by the [bot](https://github.com/Cap-go/github-bot) AKA [this](https://github.com/Cap-go/capgo/tree/main/tests_backend))

ayewo commented:
> The entire thing is still very much WIP and I am looking forward on hearing some ideas on how we can improve it

@WcaleNieWolny Thanks for the explanation. I added some tests of what I had in mind in [c1acba8](https://github.com/Cap-go/capgo/pull/399/commits/c1acba88a9d5835e903d4aa2da9fc81daa1682c7).

> Also the way you run the tests is kind of wierd. Why deno test --allow-all supabase/functions/**/index.test.ts ? Have you wrote this as a unit test and if so have you integrated this with the current testing framework (the one tiggered by the [bot](https://github.com/Cap-go/github-bot) AKA [this](https://github.com/Cap-go/capgo/tree/main/tests_backend))

OK, I have moved the tests under the `capgo/tests_backend/` folder as suggested.


PS: In order for the tests to not fail out of the box, I had to copy-paste the `jsonSchema`s used by the `/update` and `/stats` endpoints inside the tests. If Martin manages to fix the TinyBird issue in `main`, then perhaps we could remove the code duplication by deleting the copy-paste and importing it instead. I added a comment explaining my workaround in `tests_backend/zod.test.ts`.

Here's the output when the tests are run:
```bash
pn run test:zod

> capgo-app@10.182.0 test:zod /Users/mac/mnt/github/01_capgo
> deno test --allow-all tests_backend/zod.test.ts

Check file:///Users/mac/mnt/github/01_capgo/tests_backend/zod.test.ts
running 11 tests from ./tests_backend/zod.test.ts
test app_id - /updates ...
  app_id missing ... ok (11ms)
  app_id invalid #1 ... ok (5ms)
  app_id invalid #2 ... ok (7ms)
  app_id invalid #3 ... ok (8ms)
  app_id invalid #4 ... ok (7ms)
  app_id invalid #5 ... ok (7ms)
test app_id - /updates ... ok (57ms)
test device_id - /updates ...
  device_id missing ... ok (6ms)
  device_id invalid #1 ... ok (6ms)
  device_id invalid #2 ... ok (6ms)
  device_id invalid #3 ... ok (5ms)
  device_id invalid #4 ... ok (6ms)
  device_id invalid #5 ... ok (6ms)
  device_id invalid #6 ... ok (6ms)
  device_id invalid #7 ... ok (6ms)
  device_id invalid #8 ... ok (5ms)
  device_id invalid #9 ... ok (5ms)
  device_id length exceeded ... ok (7ms)
test device_id - /updates ... ok (77ms)
test version_name - /updates ...
  version_name missing ... ok (6ms)
  version_name invalid #1 ... ok (6ms)
  version_name invalid #2 ... ok (5ms)
test version_name - /updates ... ok (24ms)
test app_id and device_id - /updates ...
  app_id and device_id missing ... ok (6ms)
  app_id and device_id are invalid ... ok (5ms)
test app_id and device_id - /updates ... ok (19ms)
test app_id - /stats ...
  app_id missing ... ok (7ms)
  app_id invalid #1 ... ok (6ms)
  app_id invalid #2 ... ok (7ms)
  app_id invalid #3 ... ok (7ms)
  app_id invalid #4 ... ok (8ms)
  app_id invalid #5 ... ok (7ms)
test app_id - /stats ... ok (54ms)
test device_id - /stats ...
  device_id missing ... ok (8ms)
  device_id invalid #1 ... ok (8ms)
  device_id invalid #2 ... ok (8ms)
  device_id invalid #3 ... ok (7ms)
  device_id invalid #4 ... ok (4ms)
  device_id invalid #5 ... ok (6ms)
  device_id invalid #6 ... ok (6ms)
  device_id invalid #7 ... ok (6ms)
  device_id invalid #8 ... ok (6ms)
  device_id invalid #9 ... ok (7ms)
  device_id length exceeded ... ok (5ms)
test device_id - /stats ... ok (80ms)
test version_name - /stats ...
  version_name missing ... ok (6ms)
  version_name invalid #1 ... ok (5ms)
  version_name invalid #2 ... ok (6ms)
test version_name - /stats ... ok (25ms)
test app_id and device_id - /stats ...
  app_id and device_id missing ... ok (4ms)
  app_id and device_id are invalid ... ok (6ms)
test app_id and device_id - /stats ... ok (16ms)
test version_build - /updates ...
  version_build missing ... ok (6ms)
  version_build invalid #1 ... ok (7ms)
  version_build invalid #2 ... ok (7ms)
test version_build - /updates ... ok (27ms)
test version_os - /stats ...
  version_os missing ... ok (5ms)
  version_os invalid #1 ... ok (4ms)
  version_os invalid #2 ... ok (5ms)
test version_os - /stats ... ok (23ms)
test platform - /stats ...
  platform missing ... ok (5ms)
  platform invalid #1 ... ok (6ms)
  platform invalid #2 ... ok (5ms)
test platform - /stats ... ok (24ms)

ok | 11 passed (53 steps) | 0 failed (538ms)
```

@riderx Martin this is ready to be merged 👆.

WcaleNieWolny commented:
I am extremely confused. Why have you added tests to and yet did not integrate them with the rest of the testing framework? Adding a file to `tests_backend/` is not enough. There are already tests for the [update endpoint](https://github.com/Cap-go/capgo/blob/main/tests_backend/tests/backend/updates_test.ts) here. You could create a new file in `tests_backend/tests/backend` something like `schema_test.ts` and then again add `export function getTest()`.

See: The think is that the current CI/CD uses only this custom framework so if you want to use the code you wrote you have to start the subprocess from this custom framework. However I would much rather have [something like this](https://github.com/Cap-go/capgo/blob/6ecce9148dc50aacf54a3cbafcb31831349f59c6/tests_backend/tests/backend/updates_test.ts#L55). Reusing the existing utils + existing infrastructure would be nice

ayewo commented:
There are a number of reasons for what I did:

1. The `/stats` endpoint is being refactored by Martin so there’s a real risk that its internal workings will change soon, so it didn’t seem wise to spend time writing end-to-end tests for something that is in flux. Also my only change here was to sanitize JSON input so deploying e2e tests here would be overkill.

2. Stateful tests are harder to keep up to date so the tendency is that they’ll be ignored when anything breaks. This is why I opted for the current approach because it resembles unit testing. Only a one-line command is needed to kick off the test `pnpm run test:zod`.


3. I lean towards reuse over re-inventing the wheel. In this case I had to pick between your custom framework vs a standard framework like `ts-jest` (mentioned above).  Deno provides a built-in testing framework which is why I used it instead. It is well documented ensuring anyone other than me with some familiarity with Jest or Deno.test will be able to edit or add contributions, now or in the future.

WcaleNieWolny commented:
Those are very valid reasons for your approach and I understand why you opted out of E2E. This does not change the fact that without integration with this custom framework your tests will never be run as no one wants to clone someone's PR just to run tests.

I suggest adding a new test in the "backend" dir that will start a child process for this deno testing framework. We are already doing that for playwright so you can copy some code from there

riderx commented:
I did removed today the tiny bird integration since it doesn't work for now

riderx commented:
@WcaleNieWolny can you show the way on how to add this test to the testsuit with the child process?
I will tip you for that

riderx commented:
/tip $20 @WcaleNieWolny (review)
